src/valkey-cli.c

```diff
@@ -2398,6 +2398,34 @@ static int cliSendCommand(int argc, char **argv, long repeat) {
             fflush(stdout);
             if (config.pubsub_mode || num_expected_pubsub_push > 0) {
                 if (isPubsubPush(config.last_reply)) {
+
+                    /* Handle pubsub mode */
+                    if (config.last_reply->elements >= 3) {
+                        char *cmd = config.last_reply->element[0]->str;
+                        int count = config.last_reply->element[2]->integer;
+
+                        if (strcmp(cmd, "subscribe") == 0) {
+                            config.pubsub_unsharded_count++;
+                        } else if (strcmp(cmd, "unsubscribe") == 0) {
+                            config.pubsub_unsharded_count = count;
+                        } else if (strcmp(cmd, "psubscribe") == 0) {
+                            config.pubsub_unsharded_count++;
+                        } else if (strcmp(cmd, "punsubscribe") == 0) {
+                            config.pubsub_unsharded_count = count;
+                        } else if (strcmp(cmd, "ssubscribe") == 0) {
+                            config.pubsub_sharded_count++;
+                        } else if (strcmp(cmd, "sunsubscribe") == 0) {
+                            config.pubsub_sharded_count = count;
+                        }
+
+                        if (config.pubsub_unsharded_count == 0 && config.pubsub_sharded_count == 0) {
+                            config.pubsub_mode = 0;
+                            cliRefreshPrompt();
+                        } else {
+                            config.pubsub_mode = 1;
+                        }
```

Comment:
There is already a check that sets `pubsub_mode = 1` below. It becomes redundant with this change. We should only do this in one place.

I think it's better to do it here, but we also need to call cliRefreshPrompt and we should do it only if we're not already in pubsub mode. I suggest like this:

```suggestion
                        if (config.pubsub_unsharded_count == 0 && config.pubsub_sharded_count == 0) {
                            if (config.pubsub_mode) {
                                config.pubsub_mode = 0;
                                cliRefreshPrompt();
                            }
                        } else if (!config.pubsub_mode) {
                            config.pubsub_mode = 1;
                            cliRefreshPrompt();
                        }
```

And this code below can be deleted:

                        if (is_subscribe && !config.pubsub_mode) {
                            config.pubsub_mode = 1;
                            cliRefreshPrompt();
                        }

Reply:
I was manually testing the code and it seems that when we just use a ```unsubscribe``` command to unsubscribe to all channels the count isn't updated properly. So although we are able to unsubscribe to all channels we won't be able to automatically exit subscribe mode. I will continue to look into it.

Reply:
When unsubscribing all channels, the server sends multiple `unsubscribe` messages and the last one should have count `0`. So we should be able to detect this.

I think now that this code only detects the first unsubscribe message. The rest are received in `cliWaitForMessagesOrStdin`.

This code:

```C
        } else if (FD_ISSET(context->fd, &readfds)) {
            /* Message from the server */
            if (cliReadReply(0) != REDIS_OK) {
                cliPrintContextError();
                exit(1);
            }
            fflush(stdout);
```

It also calls `cliReadReply`. Maybe your first solution where the code was inside `cliReadReply` could actually detect the unsubscribe all channels case. We can put it back there if you want. But can we also do the check to set subscribed node to 1 here? It's good to have the code for pubsub_mode = 1 and pubsub_mode = 0 in the same place.

Reply:
Sorry for the delay. I reverted to the previous commit where the code was in `cliReadReply` and I still seem to get the same issue even though I'm sure it was working last week when I was manually testing. I put a print statement to debug and here is the output

```127.0.0.1:6379> subscribe ch1 ch2 ch3
1) "subscribe"
2) "ch1"
3) (integer) 1
1) "subscribe"
2) "ch2"
3) (integer) 2
1) "subscribe"
2) "ch3"
3) (integer) 3
127.0.0.1:6379(subscribed mode)> unsubscribe
1) "unsubscribe"
2) "ch3"
3) (integer) 2
Command: unsubscribe, Count: 2
1) "unsubscribe"
2) "ch2"
3) (integer) 1
1) "unsubscribe"
2) "ch1"
3) (integer) 0
```

So it seems to only get the reply of the first unsubscribe instead of all of them

I'm unsure how to proceed